plugins/inputs/gcs/gcs.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	return nil
+}
+
+func (gcs *GCS) fetchedMetrics(r *storage.Reader) ([]telegraf.Metric, error) {
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(r); err != nil {
+		return nil, err
+	}
+
+	return gcs.parser.Parse(buf.Bytes())
+}
+
+func (gcs *GCS) reachedThreshlod(processed int) bool {
+	return gcs.ObjectsPerIteration != 0 && processed >= gcs.ObjectsPerIteration
+}
+
+func (gcs *GCS) updateOffset(bucket *storage.BucketHandle, name string) error {
+	offsetModel := NewOffset(name)
+	marshalled, err := json.Marshal(offsetModel)
+
+	if err != nil {
+		return err
+	}
+
+	offsetKey := bucket.Object(gcs.OffsetKey)
+	writer := offsetKey.NewWriter(gcs.ctx)
+	writer.ContentType = "application/json"
+	defer writer.Close()
+
+	if _, err := writer.Write(marshalled); err != nil {
+		return err
+	}
+
+	gcs.offSet = *offsetModel
+
+	return nil
+}
+
+func (gcs *GCS) Init() error {
+	gcs.ctx = context.Background()
+	err := gcs.setUpClient()
+	if err != nil {
+		gcs.Log.Error("Could not create client", err)
+		return err
+	}
+
+	return gcs.setOffset()
+}
+
+func (gcs *GCS) setUpClient() error {
+	if endpoint, present := os.LookupEnv(emulatorHostEnv); present {
+		return gcs.setUpLocalClient(endpoint)
+	}
+
+	return gcs.setUpDefaultClient()
+}
+
+func (gcs *GCS) setUpLocalClient(endpoint string) error {
+	noAuth := option.WithoutAuthentication()
+	endpoints := option.WithEndpoint("http://" + endpoint)
+	client, err := storage.NewClient(gcs.ctx, noAuth, endpoints)
+
+	if err != nil {
+		return err
+	}
+
+	gcs.client = client
+	return nil
+}
+
+func (gcs *GCS) setUpDefaultClient() error {
+	var credentialsOption option.ClientOption
+
+	if gcs.CredentialsFile != "" {
+		credentialsOption = option.WithCredentialsFile(gcs.CredentialsFile)
+	} else {
+		creds, err := google.FindDefaultCredentials(gcs.ctx, storage.ScopeReadOnly)
+		if err != nil {
+			return fmt.Errorf(
+				"unable to find GCP Application Default Credentials: %v."+
+					"Either set ADC or provide CredentialsFile config", err)
+		}
+		credentialsOption = option.WithCredentials(creds)
+	}
+
+	client, err := storage.NewClient(gcs.ctx, credentialsOption)
+	gcs.client = client
+	return err
+}
+
+func (gcs *GCS) setOffset() error {
+	if gcs.client == nil {
+		return fmt.Errorf("Cannot set offset if client is not set")
+	}
+
+	if gcs.OffsetKey != "" {
+		gcs.OffsetKey = gcs.Prefix + gcs.OffsetKey
+	} else {
+		gcs.OffsetKey = gcs.Prefix + defaultOffSetKey
+	}
+
+	btk := gcs.client.Bucket(gcs.Bucket)
+	obj := btk.Object(gcs.OffsetKey)
+
+	var offSet OffSet
+
+	if r, err := obj.NewReader(gcs.ctx); err == nil {
+		defer gcs.closeReader(r)
+		buf := new(bytes.Buffer)
+
+		if _, err := io.Copy(buf, r); err == nil {
+			if marshalError := json.Unmarshal(buf.Bytes(), &offSet); marshalError != nil {
+				return marshalError
+			}
+		}
+	} else {
+		offSet = *NewEmptyOffset()
+	}
+
+	gcs.offSet = offSet
+
+	return nil
+}
+
+func init() {
+	inputs.Add("gcs", func() telegraf.Input {
```

Comment:
```suggestion
	inputs.Add("google_cloud_storage", func() telegraf.Input {
```
Maybe it would be better to use the full name so it is easier to identify in the config? Not sure if everyone knows "gcs" stands for google cloud storage.